apps/DesktopStreamer/MainWindow.cpp

```diff
@@ -177,205 +149,189 @@ void MainWindow::closeEvent( QCloseEvent* closeEvt )
-    else
-        _stopStreaming();
-}
-
 void MainWindow::_update()
 {
+    _frameTime.start();
+    _updateStreams();
     if( _streamButton->isChecked( ))
     {
-        _checkStream();
         _processStreamEvents();
         _shareDesktopUpdate();
+        _regulateFrameRate();
     }
     else
         _stopStreaming();
 }
 
-void MainWindow::_checkStream()
+void MainWindow::_updateStreams()
 {
-    if( _stream )
-        return;
+    const std::string& host = _getStreamHost();
 
-    QString streamHost;
-    if( _hostnameComboBox->findText( _hostnameComboBox->currentText( )) == -1 )
-        streamHost = _hostnameComboBox->currentText();
-    else
-        streamHost = _hostnameComboBox->currentData().toString();
+#ifdef DEFLECT_USE_QT5MACEXTRAS
+    const QModelIndexList windowIndices =
+        _listView->selectionModel()->selectedIndexes();
 
-    _stream.reset(
-        new deflect::Stream( _streamnameLineEdit->text().toStdString(),
-                             streamHost.toStdString( )));
-    if( !_stream->isConnected( ))
+    StreamMap streams;
+    for( QPersistentModelIndex index : windowIndices )
     {
-        _handleStreamingError( "Could not connect to host" );
-        return;
+        if( _streams.count( index ))
+        {
+            streams[ index ] = _streams[ index ];
+            continue;
+        }
+
+        const std::string name = index.isValid() ?
+            _listView->model()->data( index, Qt::DisplayRole ).
+                toString().toStdString() : std::string();
+        const std::string streamName = std::to_string( ++_streamID ) +
+                                       " " + name + " - " +
+                                      _streamnameLineEdit->text().toStdString();
+        StreamPtr stream( new Stream( *this, index, streamName, host ));
+
+        if( !stream->isConnected( ))
+        {
+            _statusbar->showMessage( std::string( "Could not connect to host " +
+                                                  host ).c_str( ));
+            continue;
+        }
+
+        if( _desktopInteractionCheckBox->isChecked( ))
+            stream->registerForEvents();
+
+        streams[ index ] = stream;
     }
+    _streams.swap( streams );
 
-#ifdef DEFLECT_USE_QT5MACEXTRAS
-    _windowIndex = _listView->currentIndex();
-    if( !_windowIndex.isValid( ))
+    if( !_streams.empty() && !_streamButton->isChecked( ))
     {
-        _handleStreamingError( "No window to stream is selected" );
+        _streamButton->setChecked( true );
+        _startStreaming();
+    }
+    _streamButton->setChecked( !windowIndices.empty( ));
+
+#else // No window list: Stream button toggles
+
+    if( !_streamButton->isChecked( ))
+    {
+        _stopStreaming();
         return;
     }
-#endif
 
-#ifdef STREAM_EVENTS_SUPPORTED
-    if( _desktopInteractionCheckBox->isChecked( ))
-        _stream->registerForEvents();
+    if( _streams.empty( ))
+    {
+        const QPersistentModelIndex index;
```

Comment:
invalid == Desktop == Index( row( 0 )). Same magic as before. All indices with row < 0 are considered desktop in the stream.